### ISSUE TITLE
CF-ServiceBrokerRequestRejected

### ISSUE THREAD

**Issue by stevegiron** [Azure/open-service-broker-azure#653]
All of a sudden we are no longer able to bind apps to services in Azure.   This app has been pushed at least 50 times and now the error out of the blue.  

PAS 2.3
AOSB 12.0
I tried to redeploy the AOSB tile and that failed too with no reason other than exit code 1.

Binding service xxxxxx to app xxxxxn in org xxxxx / space Steve as xxxx.xxxx@xxxx.com...
Unexpected Response
Response code: 502
CC code:       0
CC error code:
Request ID:    c3953ab5-e290-4d9f-73b4-16349440068a::28352509-4099-4eb9-ab3c-966738d4500f
Description:   {
  "description": "The service broker rejected the request to https://open-service-broker-azure.apps.pcf.nonprod.xxxxxx.com/v2/service_instances/77a087e5-2f26-4f8a-92c3-1ddb10dc03da/service_bindings/5691335f-410e-4635-831a-52930b5d3228. Status Code: 400 Bad Request, Body: {}",
  "error_code": "CF-ServiceBrokerRequestRejected",
  "code": 10001,
  "http": {
    "uri": "https://open-service-broker-azure.apps.pcf.nonprod.xxxxxx.com/v2/service_instances/77a087e5-2f26-4f8a-92c3-1ddb10dc03da/service_bindings/5691335f-410e-4635-831a-52930b5d3228",
    "method": "PUT",
    "status": 400
  }
}

**Comment by norshtein** [comment]
Could you check the running log of the OSBA? You may first run command `cf logs <OSBA_APP_NAME>` and then try to bind a service instance to your app.

**Comment by stevegiron** [comment]
There is no return when I do the '''cf logs'' command and then bind. I still get the same reject but nothing shows up on the log window.

**Comment by norshtein** [comment]
That's odd... Could you double check you don't have several OSBA apps, and you haven't referred different app in `cf logs` and `cf create-service-broker`?

If the above checking is correct, could you run `cf logs <OSBA_APP_NAME>` and then try to create any service instance? And observe will there be any log?

**Comment by stevegiron** [comment]
@zhongyi-zhang Yes, I used two terminals. I've opened a case with MS and have sent them CC logs.

**Comment by stevegiron** [comment]
Yes we're using Basic Tier Azure Redis and the cache looks good.

**Comment by bingosummer** [comment]
@stevegiron 
There's a button "Console" in the overview of the redis cache on Azure Portal. You can click "Console" to run `KEYS *`. If the result is empty, then the data is lost.

**Comment by stevegiron** [comment]
``` >KEYS *
KEYS command is unsupported. Sending "SCAN 0 COUNT 1000 MATCH *" instead ```

**Comment by stevegiron** [comment]
I tried the KEYS command on another fully functional in production Redis cache with the same result.

**Comment by zhongyi-zhang** [comment]
@stevegiron there is no more rows? If so, we can confirm that was a data loss :(. OSBA can't help restore them.
See https://azure.microsoft.com/en-us/pricing/details/cache/. The basic cache doesn’t support replication, failover, or data persistence. Some related topics also can be found in stackoverflow: https://stackoverflow.com/questions/51208098/azure-redis-cache-service-lost-all-data.

**Comment by zhongyi-zhang** [comment]
Here is an example what looks good.

I have an `azure-postgresql-9-6` instance provisioned:
```
name   service                plan    bound apps   last operation
ap     azure-postgresql-9-6   basic                create succeeded
```
In the database, we can see something like:

![image](https://user-images.githubusercontent.com/13326941/50262021-19e46400-044a-11e9-9456-7dc584c8f06c.png)

You can confirm the data loss by this way.

When Redis data loss happened, OSBA can't manage the resources any longer.
Now the only solution is to purge the orphan service instances by `cf purge-service-instance <instance-name>` and recreate them. (Don't forget to also delete the resources manully on Azure Portal or Azure CLI as they are still costing.)

And please never use Basic Tier Redis in non-test environment...

**Comment by stevegiron** [comment]
Thanks Zhonghy. I do have more diagnostic information I'll share tomorrow, however your suspicion of a Redis issue is correct.

**Comment by stevegiron** [comment]
It was determined that a likely cause of the Redis database corruption/data loss due to our OnPrem to Azure VPN tunnel experiencing an intermittent outage over a 4 day period.

We were able to redeploy AOSB and create new services and bind to them. We were not able to bind to any existing services that were deployed via the AOSB prior to the VPN tunnel outage.

It would be nice to see the ability to configure backups and have a higher level of availability for Redis, especially for production environments.  Because the AOSB is not supported at an enterprise level, our management has chosen to abandon the AOSB until it is a supported product.

**Comment by skibum55** [comment]
Working with @stevegiron we found a debug message that wasn't bubbling up.  It would have been nice to see it along with the bind failure.
<img width="1648" alt="screen shot 2018-12-20 at 5 08 01 pm" src="https://user-images.githubusercontent.com/3813241/50317684-1799fe00-047a-11e9-873a-bec712c02063.png">

**Comment by stevegiron** [comment]
We were able to use our Enterprise Microsoft resources to look at this as well.  Here is information we just received on this. ``` our Basic Redis does not have SLA. This is because the backend of our Redis product is Cloud Service. And the Basic Redis only has one Cloud Service instance. Therefore, during the OS pathcing or other issues on the Cloud Service, all the data in Basic Redis will be purged. Therefore, in a production environment, we highly suggest that our customers use at least Standard Pricing tier Redis.

They found both redis instances have been patched at 2018-12-13 19:00:25 UTC. During the patching, the instances of the Redis were rebooted, so the data disappeared. ```

**Comment by stevegiron** [comment]
We were able to get MS Enterprise support to take a look at Redis from their end.  Here is the commentary from MS.  
``` Our Basic Redis does not have SLA. This is because the backend of our Redis product is Cloud Service. And the Basic Redis only has one Cloud Service instance. Therefore, during the OS pathcing or other issues on the Cloud Service, all the data in Basic Redis will be purged. Therefore, in a production environment, we highly suggest that our customers use at least Standard Pricing tier Redis.

They found both redis instances have been patched at 2018-12-13 19:00:25 UTC. During the patching, the instances of the Redis were rebooted, so the data disappeared. ```

**Comment by zhongyi-zhang** [comment]
@skibum55 yes, that's expected behavior to a data loss in the broker store (the Redis).
